arangod/Aql/OptimizerRules.cpp

```diff
@@ -3151,27 +3150,30 @@ struct SortToIndexNode final
         condition->normalize(_plan);
         TRI_ASSERT(usedIndexes.size() == 1);
         IndexIteratorOptions opts;
-        opts.ascending = sortCondition.isAscending();
+        TRI_ASSERT(!sortCondition.isEmpty());
+        opts.ascending = sortCondition.sortFields()[0].asc;
         opts.useCache = false;
-        auto n = _plan->createNode<IndexNode>(
+        auto indexNode = _plan->createNode<IndexNode>(
             _plan, _plan->nextId(), enumerateCollectionNode->collection(),
             outVariable, usedIndexes,
             false,  // here we could always assume false as there is no lookup
                     // condition here
             std::move(condition), opts);
 
-        enumerateCollectionNode->CollectionAccessingNode::cloneInto(*n);
-        enumerateCollectionNode->DocumentProducingNode::cloneInto(_plan, *n);
+        enumerateCollectionNode->CollectionAccessingNode::cloneInto(*indexNode);
+        enumerateCollectionNode->DocumentProducingNode::cloneInto(_plan,
+                                                                  *indexNode);
 
-        _plan->replaceNode(enumerateCollectionNode, n);
+        _plan->replaceNode(enumerateCollectionNode, indexNode);
         _modified = true;
 
         if (coveredAttributes == sortCondition.numAttributes()) {
-          // if the index covers the complete sort condition, we can also remove
-          // the sort node
-          n->needsGatherNodeSort(
-              makeIndexSortElements(sortCondition, outVariable));
-          _plan->unlinkNode(_plan->getNodeById(_sortNode->id()));
+          deleteSortNode(indexNode, sortCondition);
+        } else if (usedIndexes.size() == 1 &&
+                   usedIndexes[0]->type() ==
+                       Index::IndexType::TRI_IDX_TYPE_PERSISTENT_INDEX) {
+          auto sortNode = _plan->getNodeById(_sortNode->id());
+          ((SortNode*)sortNode)->setGroupedElements(coveredAttributes);
```

Comment:
```suggestion
          _sortNode->setGroupedElements(coveredAttributes);
```
Probably I'm missing something here. Anyways, instead of the old school c-style cast, please use `ExecutionNode::castTo<SortNode*>()`. This does a `dynamic_cast` if required and a `static_cast` is possible.